src/main/java/forestry/factory/recipes/RecipeSerializers.java

```diff
@@ -0,0 +1,88 @@
+/*******************************************************************************
+ * Copyright (c) 2011-2014 SirSengir.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the GNU Lesser Public License v3
+ * which accompanies this distribution, and is available at
+ * http://www.gnu.org/licenses/lgpl-3.0.txt
+ *
+ * Various Contributors including, but not limited to:
+ * SirSengir (original work), CovertJaguar, Player, Binnie, MysteriousAges
+ ******************************************************************************/
+package forestry.factory.recipes;
+
+import com.google.gson.JsonObject;
+
+import java.util.function.BiConsumer;
+import java.util.function.Function;
+
+import net.minecraft.fluid.Fluid;
+import net.minecraft.item.ItemStack;
+import net.minecraft.nbt.CompoundNBT;
+import net.minecraft.nbt.NBTDynamicOps;
+import net.minecraft.network.PacketBuffer;
+import net.minecraft.util.JSONUtils;
+import net.minecraft.util.NonNullList;
+import net.minecraft.util.ResourceLocation;
+
+import com.mojang.serialization.Dynamic;
+import com.mojang.serialization.JsonOps;
+
+import net.minecraftforge.fluids.FluidStack;
+import net.minecraftforge.registries.ForgeRegistries;
+
+public class RecipeSerializers {
+
+	static <E> void write(PacketBuffer buffer, NonNullList<E> list, BiConsumer<PacketBuffer, E> consumer) {
+		buffer.writeVarInt(list.size());
+
+		for (E e : list) {
+			consumer.accept(buffer, e);
+		}
+	}
+
+	static <E> NonNullList<E> read(PacketBuffer buffer, Function<PacketBuffer, E> reader) {
+		NonNullList<E> list = NonNullList.create();
+		int size = buffer.readVarInt();
+
+		for (int i = 0; i < size; i++) {
+			list.add(reader.apply(buffer));
+		}
+
+		return list;
+	}
+
+	public static FluidStack deserializeFluid(JsonObject object) {
+		ResourceLocation fluidName = new ResourceLocation(JSONUtils.getString(object, "fluid"));
+		Fluid fluid = ForgeRegistries.FLUIDS.getValue(fluidName);
+		FluidStack stack = new FluidStack(fluid, JSONUtils.getInt(object, "amount", 0));
+
+		if (object.has("tag")) {
+			stack.setTag((CompoundNBT) Dynamic.convert(JsonOps.INSTANCE, NBTDynamicOps.INSTANCE, object.get("tag")).copy());
+		}
+
+		return stack;
+	}
+
+	public static JsonObject serializeFluid(FluidStack fluid) {
+		JsonObject object = new JsonObject();
+		object.addProperty("fluid", fluid.getFluid().getRegistryName().toString());
+		object.addProperty("amount", fluid.getAmount());
+
+		if (fluid.hasTag()) {
+			object.add("Tag", Dynamic.convert(NBTDynamicOps.INSTANCE, JsonOps.INSTANCE, fluid.getTag()));
+		}
+
+		return object;
+	}
+
+	public static JsonObject item(ItemStack stack) {
```

Comment:
Is there no vanilla / forge method for this?

Reply:
Apparently not. I first thought of using Dynamics to convert the NBT item to a JSON object but I don't think ShapedRecipe.deserializeItem would work with it